### ISSUE TITLE
Dependency types for Dockerfile

### ISSUE THREAD

**Issue by mxey** [renovatebot/renovate#2364]
**What would you like Renovate to be able to do?**

In a multi-stage Docker build, some of the source images might only used as build tools, while the last one becomes the base of the resulting image. I'd like to be able to tell those apart, similar to `dependencies` and `devDependencies`, so I can automerge or pin just one of them, for example

**Describe the solution you'd like**

I see multiple options:

1. Report all but the last image as `devDependencies` or `buildImage` or similar
2. Use the alias names of the stages as the type

Option 1 would be more generic and probably fit most use cases. 

**Additional context**

* https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds
* https://docs.docker.com/develop/develop-images/multistage-build/#name-your-build-stages

**Comment by rarkins** [comment]
First option sounds fine to me, it’s just a matter of picking names that make sense. Eg “intermediate” and “final”, for example

**Comment by rarkins** [comment]
Let's use "stage" and "final" as the two dep types. Any FROM ... AS or COPY FROM should have depType=stage while the final FROM without AS should be depType=final. I think it's possible syntactically to also have 'AS foo' on the final FROM so it would be nice if we detected that as final too

**Comment by Dmitry9** [comment]
So it would be like:
```json
{
  "docker": {
    "major": {
      "enabled": true
    },
  "multi-stage build" : {
     "enabled": true,
     "intermediate image name alias to omit" : "stage"
     }
  }
}
```
So ` "intermediate image name alias to update" : "builder" `    can be ommited as default one e.g. renovate will look for string `builder` inside string `COPY --from=builder /.....`

But if you have `COPY --from=stage /.....` then above json will do the trick. 
If "multi-stage build" ["enabled] === false then any value of "intermediate image name alias to omit" should be ignored.
Then change extract function logic so it inside this content:
```docker
FROM golang:1.7.3 as builder
WORKDIR /go/src/github.com/alexellis/href-counter/
RUN go get -d -v golang.org/x/net/html  
COPY app.go    .
RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o app .

FROM alpine:3.01.01  
RUN apk --no-cache add ca-certificates
WORKDIR /root/
COPY --from=builder /go/src/github.com/alexellis/href-counter/app .
CMD ["./app"]  
```
will extract only 3.1.1 (and ignore 1.7.3) 
then -> check semver -> lookup(docker registry -> alpine)-> udate (how to update(package rules[only patch, only minor ....]))

**Comment by Dmitry9** [comment]
also packageRules.managers should contain 'Dockerfile' I think in this case.

**Comment by rarkins** [comment]
@Dmitry9 actually it's not this complicated. We just want the dockerfile extract function to include the field `depType` whenever it extracts dependencies.

You can see here an example snapshot from our tests: https://github.com/renovatebot/renovate/blob/6837556e10332d7e6c60b57069c0ceab4604ac59/test/manager/dockerfile/__snapshots__/extract.spec.js.snap#L31-L57

In this case, we'd want the first one to include `"depType": "stage"` and the second one to include `"depType": "final"`

This would be the only change - only the manager/dockerfile/extract.js file should change and its associated tests and snapshots.

Is that clearer?

**Comment by Dmitry9** [comment]
@rarkins thank you, yes, it is clearer now. I am putting this issue in progress status.

**Comment by Dmitry9** [comment]
@rarkins I see my changes are merged. Can you please drop a high-level overview of what else needs to be done to close this issue?

**Comment by rarkins** [comment]
Yes, this can be closed. The PR needed to have `Closes #2364` and not `Closes # 2364` in its body for GitHub to link and autoclose it.
